compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUMergeParallelOps.cpp

```diff
@@ -0,0 +1,72 @@
+// Copyright 2023 The IREE Authors
```

Comment:
If I am not mistaken this is doing rematerialization of some operation. I am not sure this is the right way to go here. It is sidestepping what is really a bufferization issue.

Reply:
Correct it does rematerialization. We need it independently of bufferization improvements, for performance we don't want to go back to memory as those kernels are highly memory bound and bufferization won't always be able to store the intermediate result in output buffer. (for instance if the intermediate result has a different type than the output buffer)
But overall in order to get the perf we want for softmax kind of ops we need to do the rematerialization.

Reply:
This is the worrying part. This seems to be an example where the something needed specifically for softmax is being done by default. Rematerialization needs to be done carefully. It not only increases compute but in general also increases bandwidth requirements (you need to read all the operands again to perform the computation). In very specific case it all works out and softmax seems to be that specific case.

Reply:
> This means we have to postpone the work on enabling aggressive-fusion until we have a solution as it would either crash the compiler or regress performance compared to not doing the fusion.

The current implementation of aggressive fusion is purposely pushing the limit of what we can do default heuristics. So there might be intermediate things we can enable, but most of it actually works by default (except for softmax for the reasons here). Have been chatting with Matthias about the issue with bufferization. There was something that was done for transform dialect path that maybe can be used. There are some ad-hoc patterns we can add to convert this to destination passing style which solves the problem.
I am fine with ad-hoc intermediate solutions as long as it is scoped properly. This is a pass that is being added by default.. If we are looking at ad-hoc solution, then fixing bufferization for this case is less ad-hoc than doing something different for CUDA backend (as in we would do the same ad-hoc thing on all backends which is the only reason to prefer that). 

> Not directly related but I'm wondering if we actually want to fuse in case we think we should have a large temporary result as we won't usually save on any bandwidth and we are at risk of not having enough memory to store the intermediate result (as mentioned I don't think bufferization fix would solve it all). So in a way the real heuristic decision was already decided when we do the fusion.

It depends. If we are propagating the results through registers (or through small buffers) then that is the only rationale for fusing these things. I would drop this fusion at a drop of the hat, but since this has been a topic of discussion for a long time now I pushed the default heuristics to what I consider its limit to account for it.

Reply:
> I am fine with ad-hoc intermediate solutions as long as it is scoped properly. This is a pass that is being added by default.. If we are looking at ad-hoc solution, then fixing bufferization for this case is less ad-hoc than doing something different for CUDA backend (as in we would do the same ad-hoc thing on all backends which is the only reason to prefer that).

It's added by default only for the group reduction pipeline, it doesn't touch any other cases. In my opinion the cases where we want to fuse but prefer to store the intermediate result in memory to avoid recomputation is going to be the less common path. We could build an heuristic for it but I would rather not do that without more data. Storing back in the output buffer is only going to be a workaround in my opinion as this is rarely what we would want to do.

> It depends. If we are propagating the results through registers (or through small buffers) then that is the only rationale for fusing these things.

If we keep it in register then CSE would undo the rematerialization out of the box. Small buffers is the case where it would help and where there is an interesting tradeoffs, This is the case where it could go in a temp alloc.


> I would drop this fusion at a drop of the hat, but since this has been a topic of discussion for a long time now I pushed the default heuristics to what I consider its limit to account for it.

I'm happy completely dropping this path if we have a clear actionable plan to get back those performance without those default heuristics. It seems like a case where making the right choice requires making a "good" coherent decision of fusion and codegen based on target characteristics.